File: client/src/main/java/pro/panopticon/client/sensor/impl/FileHandlesSensor.java
package pro.panopticon.client.sensor.impl;

import com.amazonaws.services.cloudwatch.model.StandardUnit;
import pro.panopticon.client.model.Measurement;
import pro.panopticon.client.sensor.Sensor;
import pro.panopticon.client.util.SystemStatus;

import java.util.ArrayList;
import java.util.List;

public class FileHandlesSensor implements Sensor {

    private final long warnAfter;
    private final long errorAfter;

    public FileHandlesSensor(long warnAfter, long errorAfter) {
        this.warnAfter = warnAfter;
        this.errorAfter = errorAfter;
    }

    @Override
    public List<Measurement> measure() {
        SystemStatus s = new SystemStatus();

        List<Measurement> measurements = new ArrayList<>();

        long open = s.openFileHandles();
        long max = s.maxFileHandles();
        double percent = ((double) open / (double) max) * 100;
        String displayValue = String.format("%s of %s filehandles used (%.2f%%)", open, max, percent);

        measurements.add(new Measurement("filehandles", statusFromOpenFileHandles(open), displayValue, new Measurement.CloudwatchValue(percent, StandardUnit.Percent)));

        return measurements;
    }

    private String statusFromOpenFileHandles(long open) {
        if (open >= errorAfter) {
            return "ERROR";
        } else if (open >= warnAfter) {
            return "WARN";
        } else {
            return "INFO";
        }
    }

}

File: client/src/main/java/pro/panopticon/client/sensor/impl/FileHandlesSensor.kt
package pro.panopticon.client.sensor.impl

import com.amazonaws.services.cloudwatch.model.StandardUnit
import pro.panopticon.client.model.Measurement
import pro.panopticon.client.sensor.Sensor
import pro.panopticon.client.util.SystemStatus

class FileHandlesSensor(private val warnAfter: Long, private val errorAfter: Long) : Sensor {
    override fun measure(): List<Measurement> {
        val s = SystemStatus()
        val open = s.openFileHandles()
        val max = s.maxFileHandles()
        val percent = open.toDouble() / max.toDouble() * 100
        val displayValue = String.format("%s of %s filehandles used (%.2f%%)", open, max, percent)

        return listOf(
            Measurement(
                key = "filehandles",
                status = statusFromOpenFileHandles(open),
                displayValue = displayValue,
                cloudwatchValue = Measurement.CloudwatchValue(percent, StandardUnit.Percent),
                description = ""
            )
        )
    }

    private fun statusFromOpenFileHandles(open: Long): String {
        return when {
            open >= errorAfter -> "ERROR"
            open >= warnAfter -> "WARN"
            else -> "INFO"
        }
    }
}
